File: Doubly-Linked-List/dll.go
package main

import (
  "fmt"
)

type Node struct {
  data int
  prev *Node
  next *Node
}

func main() {
  var head *Node

  for i := 1; i <= 5; i++ {
    head = insertAtBeginning(head, i)
  }
  printListReverse(head)
}

func insertAtBeginning(head *Node, data int) *Node {
  // create a new node
  newNode := &Node{}
  newNode.data = data

  // If the list is empty
  if head == nil {
    head = newNode
    return head
  }

  newNode.next = head
  head.prev = newNode
  head = newNode
  return head
}

func printList(head *Node) {
  for current := head; current != nil; current = current.next {
    fmt.Printf("%d ", current.data)
  }
}

func printListReverse(head *Node) {

  if head == nil {
    return
  }

  var reversePointer *Node

  // Traverse to the last node of the doubly linked list
  for current := head; current != nil; current = current.next {
    reversePointer = current
  }

  // Traverse back to the first node of the doubly linked list
  for ; reversePointer != nil; reversePointer = reversePointer.prev {
    fmt.Printf("%d ", reversePointer.data)
  }
}

File: Doubly-Linked-List/dll.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class DLL:
    def __init__(self):
        self._head = None

    def insertAtBeginning(self, data):
        # create a new node
        new_node = Node(data)

        # check if the list is empty
        if self._head is None:
            self._head = new_node
            return

        new_node.next = self._head
        self._head.prev = new_node
        self._head = new_node

    def printList(self):
        print("Printing List...")
        current = self._head
        while current:
            print(current.data)
            current = current.next

    def printListReverse(self):
        if self._head is None:
            return

        print("Printing List in reverse...")
        current = self._head
        while current.next:
            current = current.next

        while current:
            print(current.data)
            current = current.prev


if __name__ == "__main__":
    dllist = DLL()
    for i in range(1, 6):
        dllist.insertAtBeginning(i)
    dllist.printList() # 5 4 3 2 1
    dllist.printListReverse() # 1 2 3 4 5
